server/src/Application/Services/FlashcardImporter/FlashcardImporter.cs

```diff
@@ -0,0 +1,34 @@
+using FluentValidation;
+using StudyZen.Application.Dtos;
+using StudyZen.Application.Extensions;
+using System.Collections.Concurrent;
+
+namespace StudyZen.Application.Services;
+
+public sealed class FlashcardImporter : IFlashcardImporter
+{
+    public IReadOnlyCollection<CreateFlashcardDto> ImportFlashcardsFromCsv(Stream stream, int flashcardSetId)
+    {
+        var lines = stream.ReadLines();
+        var importedFlashcards = new BlockingCollection<CreateFlashcardDto>();
+
+        Parallel.ForEach(lines, line =>
+        {
+            var values = line.Split(',');
+            if (values.Length != 2)
+            {
+                throw new ValidationException($"CSV file must only contain flashcard front and back values. Invalid values: {string.Join(',', values)}");
```

Comment:
I would suggest adding a file validator to check, for example, file's size and type, because now I can try to import pdf, but still get the general exception about CSV not containing  front and back values. 
`    public class FlashcardsFileValidator : AbstractValidator<IFormFile>
    {
        public FlashcardsFileValidator()
        {
            RuleFor(f => f.Length).NotNull().LessThanOrEqualTo(5000000) // or whatever file size in bytes
                .WithMessage("File size is larger than allowed");

            RuleFor(f=> f.ContentType).NotNull().Must(f => f.Equals("text/csv"))
                .WithMessage("Only .csv files are allowed");
        }
    }`

Reply:
`
 public class FlashcardsFileValidator : AbstractValidator<IFormFile>
    {
        public FlashcardsFileValidator()
        {
            RuleFor(f => f.Length).NotNull().LessThanOrEqualTo(5000000) // or whatever file size in bytes
                .WithMessage("File size is larger than allowed");

            RuleFor(f=> f.ContentType).NotNull().Must(f => f.Equals("text/csv"))
                .WithMessage("Only .csv files are allowed");
        }
    }
`

Reply:
public class FlashcardsFileValidator : AbstractValidator<IFormFile>
    {
        public FlashcardsFileValidator()
        {
            RuleFor(f => f.Length).NotNull().LessThanOrEqualTo(5000000) // or whatever file size in bytes
                .WithMessage("File size is larger than allowed");

            RuleFor(f => f.ContentType).NotNull().Must(f => f.Equals("text/csv"))
                .WithMessage("Only .csv files are allowed");
        }
    }

Reply:
a possible example

Reply:
Agreed. Just some thoughts: since IFormFile comes from Microsoft.AspNetCore.Http, the file should probably be kept in Api project (no reference in Application project).

Maybe could pass the ContentType string to the service, which could then validate it.

Reply:
Oh u know what, we could use regex here and get an additional requirement:

e.g.: https://stackoverflow.com/questions/374930/validating-file-types-by-regular-expression

Reply:
Good idea